### PR TITLE
Dm 2360

### PR DESCRIPTION
**Pull request by JoshingYou1**
### JIRA issue link
https://agile6.atlassian.net/browse/DM-2360

## Description - what does this code do?
Adds introduction and map to VISN show page.

## Testing done - how did you test it/steps on how can another person can test it 
**_Initial Setup_**
1. run `rails db:migrate`.
2. run `rails visns:create_visn_liaisons_and_transfer_data`.
3. One VaFacility did not have lat/lng coordinates, so open up a rails console, run `VaFacility.destroy_all` and then run `rails va_facilities:create_va_facilities_and_transfer_data`.

_**Introduction**_
1. Visit the show page of any VISN (ex: `/visns/2`).
2. Make sure you see the intro text on the left side of the page that contains the number of facilities within that visn, the states within that visn, how many practices were created and adopted within that visn, and the facility types and counts within that visn.
3. Make sure you see the primary visn liaison info on the right side of the page, and that the mailto link is working properly.
4. Click on the `Jump to list of VA facilities` link and make sure you're sent to `/facilities` AKA the `va_facilities` index page.

_**Map**_
1. Select one of the facility markers on the map and remember where it is located, the name of the facility, and the practices created/adopted counts.
2. Open a new browser window and create/update a practices originating facility and/or create/update an adoption with the facility from the previous step.
3. Go back to the VISN show page and refresh it.
4. Click on the same facility marker from step 1 and make sure the counts were updated accordingly.
5. Click on the name of the facility within the marker modal and make sure the link opens a new window and takes you to the show page of that facility (ex: `/facilities/common-name`).

## Screenshots, Gifs, Videos from application (if applicable)
Screen Shot 2021-04-06 at 4 27 16 PM
Screen Shot 2021-04-06 at 4 28 33 PM

## Link to mock-ups/mock ups (image file if you have it) (if applicable)
https://www.figma.com/file/fNXjZzLUS39bKmjdfDWyxy/My-VISN%2FVAMC?node-id=1%3A171

## Acceptance criteria
- Build 18 pages, one for each VISN
- VISN Number and name in header
- Text auto-populates number of facilities, states VISN is in, # created, # adopted in left hand header
- Liaison info, role, phone #, email, where they’re located in righthand header 
- VISN page has map with all facilities in the VISN
- Jump button link to list of VAMCs
- Configure VISN route

## Definition of done
- [X] Unit tests written (if applicable)
- [X] e2e/accessibility tests written (if applicable)
- [X] Events are logged appropriately
- [X] Documentation has been updated, if applicable
- [X] A link has been provided to the originating JIRA issue
- [X] No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs

### CODE REVIEW

**Review by emyl3 — COMMENTED** [review]
Functionality-wise it looked good to me Josh! I just left some comments/questions! Let me know if you need any clarification :)

File: app/views/visns/maps/_show_map.html.erb [hunk 1]
```diff
@@ -0,0 +1,17 @@
+<% if ENV['GOOGLE_API_KEY'].present? %>
+  <% provide :head_tags do %>
+    <%= javascript_tag 'data-turbolinks-track': 'reload' do %>
+      var visnVaFacilityMapData = <%= raw @visn_va_facility_markers.to_json %>;
+      var defaultVisnVaFacilityMarkerIcon = '<%= image_path('visn-va-facility-map-marker-default.svg') %>';
+      var hoverVisnVaFacilityMarkerIcon = '<%= image_path('visn-va-facility-map-marker-hover.svg') %>';
+      var selectedVisnVaFacilityMarkerIcon = '<%= image_path('visn-va-facility-map-marker-selected.svg') %>';
+    <% end %>
+    <%= javascript_include_tag 'visns/maps', 'data-turbolinks-track': 'reload' %>
+    <%#= javascript_include_tag 'visns/_map_utilities', 'data-turbolinks-track': 'reload' %>
```
[the remark this answers is not in the document]
Comment (JoshingYou1) [reply app/views/visns/maps/_show_map.html.erb:10]: removed

File: app/views/visns/maps/_index_map.html.erb [hunk 2]
```diff
@@ -0,0 +1,17 @@
+<% if ENV['GOOGLE_API_KEY'].present? %>
+  <% provide :head_tags do %>
+    <%= javascript_tag 'data-turbolinks-track': 'reload' do %>
+      var visnMapData = <%= raw @visn_markers.to_json %>;
+      var defaultVisnVaFacilityMarkerIcon = '<%= image_path('visn-va-facility-map-marker-default.svg') %>';
+      var hoverVisnVaFacilityMarkerIcon = '<%= image_path('visn-va-facility-map-marker-hover.svg') %>';
+      var selectedVisnVaFacilityMarkerIcon = '<%= image_path('visn-va-facility-map-marker-selected.svg') %>';
+    <% end %>
+    <%= javascript_include_tag 'visns/maps', 'data-turbolinks-track': 'reload' %>
+    <%#= javascript_include_tag 'visns/_map_utilities', 'data-turbolinks-track': 'reload' %>
```
[the remark this answers is not in the document]
Comment (JoshingYou1) [reply app/views/visns/maps/_index_map.html.erb:10]: removed

File: app/helpers/visns_helper.rb [hunk 3]
```diff
@@ -1,35 +1,140 @@
 module VisnsHelper
-  def get_adopted_practices_count_by_visn(visn)
-    practices = Practice.where(approved: true, published: true, enabled: true)
+  include StatesHelper
+
+  def published_enabled_approved_practices
+    Practice.published_enabled_approved
```
[the remark this answers is not in the document]
Comment (JoshingYou1) [reply app/helpers/visns_helper.rb:5]: removed

File: app/helpers/visns_helper.rb [hunk 4]
```diff
@@ -1,35 +1,140 @@
...
+  end
+
+  def visn_va_facilities(visn)
+    VaFacility.cached_va_facilities.where(visn: visn)
```
[the remark this answers is not in the document]
Comment (JoshingYou1) [reply app/helpers/visns_helper.rb:9]: Moved this to the model.

File: app/helpers/va_facilities_helper.rb [hunk 5]
```diff
@@ -0,0 +1,17 @@
+module VaFacilitiesHelper
+  include VisnsHelper
+
+  def get_practices_created_count_by_va_facility(va_facility)
+    Practice.joins(:practice_origin_facilities).where(published: true, enabled: true, approved: true, practice_origin_facilities: { facility_id: va_facility.station_number.to_s }).count
```
[the remark this answers is not in the document]
Comment (JoshingYou1) [reply app/helpers/va_facilities_helper.rb:5]: Well, this and the other method in this helper file gather data based on the `va_facility`, and I'm only using them in the view, so I think this is acceptable. If you object, let me know :blush:

File: app/helpers/visns_helper.rb [hunk 6]
```diff
@@ -1,35 +1,140 @@
...
+    end
+
+    location_list
+  end
+
+  def facility_type_count(facility_type_array, facility_type)
+    facility_type_array.select { |type| type === facility_type }.count
+  end
+
+  def get_facility_types_and_counts_by_visn(visn)
```
[the remark this answers is not in the document]
Comment (JoshingYou1) [reply app/helpers/visns_helper.rb:91]: refactored

File: app/helpers/va_facilities_helper.rb (hunk 5 above)
[the remark this answers is not in the document]
Comment (emyl3) [reply app/helpers/va_facilities_helper.rb:4]: I'm still sometimes unsure of where some code should go in Rails... I have no reason to object at the moment... if I ever stumble across anything in my readings I will let you know 😆

File: app/helpers/visns_helper.rb [hunk 7]
```diff
@@ -1,35 +1,144 @@
...
+      facility_type_count(visn_facility_types_arr, 'Other Outpatient Services (OOS)'),
+      facility_type_count(visn_facility_types_arr, 'Primary Care CBOC'),
+      facility_type_count(visn_facility_types_arr, 'Residential Care Site (MH RRTP/DRRTP) (Stand-Alone)'),
+      facility_type_count(visn_facility_types_arr, 'Unclassified'),
+      facility_type_count(visn_facility_types_arr, 'VA Medical Center (VAMC)')
+    ]
+  end
+
+  def get_facility_type_text_by_visn(visn)
+    health_care_center_count = facility_type_counts_by_visn(visn)[0]
```
Comment (emyl3) [inline app/helpers/visns_helper.rb:106]: could we just set the value of `facility_type_counts_by_visn` to a variable and then call that variable when we are assigning the counts so we don't have to call the same method multiple times?

**Review by emyl3 — COMMENTED** [review]
one small thing! 😅

File: app/helpers/visns_helper.rb (hunk 7 above)
Comment (JoshingYou1) [reply app/helpers/visns_helper.rb]: refactored

**Review by bradjohnson92008 — APPROVED** [review]
I retested with your latest changes - Josh - everything worked as expected.  Great job Josh!.

**Review by emyl3 — APPROVED** [review]

**Pull request merged**